extensions/amp-date-picker/0.1/amp-date-picker.js

```diff
@@ -210,16 +270,120 @@ class AmpDatePicker extends AMP.BaseElement {
     /** @private {?function(?):?} */
     this.templateThen_ = null;
 
+    /** @private {?Element} */
+    this.startDateField_ = null;
+
+    /** @private {?Element} */
+    this.endDateField_ = null;
+
+    /** @private {?Element} */
+    this.dateField_ = null;
+
+    /** @private */
+    this.renderedTemplates_ = map();
+
+    /** @private {?function()} */
+    this.templatesReadyResolver_ = null;
+
+    /** @private {!Promise} */
+    this.templatesReadyPromise_ = new Promise(resolve => {
+      this.templatesReadyResolver_ = resolve;
+    });
+
+    /** @private @const {!Array<!UnlistenDef>} */
+    this.unlisteners_ = [];
+
+    const initialState = (this.mode_ == DatePickerMode.OVERLAY ?
+      DatePickerState.OVERLAY_CLOSED :
+      DatePickerState.STATIC);
     /** @private @const */
-    this.registerAction_ = this.registerAction.bind(this);
+    this.stateMachine_ = new FiniteStateMachine(initialState);
+    this.setupStateMachine_(this.stateMachine_);
 
     const locale = this.element.getAttribute('locale') || DEFAULT_LOCALE;
     this.moment_.locale(locale);
   }
 
   /** @override */
   isLayoutSupported(layout) {
-    return layout == Layout.CONTAINER;
+    return layout == Layout.CONTAINER || layout == Layout.NODISPLAY;
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.setupTemplates_();
+    // Make sure it's rendered and measured properly
+    return this.render(this.state_);
+  }
+
+  /** @override */
+  resumeCallback() {
+    this.setupListeners_();
+    this.setupTemplates_();
+  }
+
+  /** @override */
+  pauseCallback() {
+    this.cleanupListeners_();
```

Comment:
Is handling `pause` and `resume` needed or just here for performance? If just for perf, I suggest moving it to `unlayoutCallback` instead of bothering with pause/resume. ( I am not too confident all components that handle their children themselves managel pause/resume properly)

Reply:
I thought it would be beneficial to stop listening to events when the document would be swiped away from in the viewer. I can't stop listening in unlayout callback because of the lightbox use-case. `amp-lightbox` calls unlayouts its children when the lightbox closes, and I want the date picker to continue listening even after a lightbox containing it closes. Do you see a better approach than this?

Reply:
mmm, lightbox calls `pause` on `close` not `unlayout`. You may need to override the `unlayoutOnPause` for the component? Ideally you have this code in `unlayout` instead of `pause`.